Use server-side processing for packages table

replaceafill opened the pull request:
This PR changes the implementation of the two tables that display packages in the Storage Service. It adds a new view (`/packages_ajax`) that uses the [server-side processing options](http://legacy.datatables.net/usage/server-side) of the DataTable library to retrieve small batches of packages through AJAX requests instead of the full list of packages.

It also addresses the following changes proposed by Sara and the rest of the team in https://github.com/archivematica/Issues/issues/676#issuecomment-497077314 and below:

* Deleted the Description column which was set always to None. I think this is caused by not having the `description` field included in `storage_service.locations.api.PackageResource.Meta.fields`
* Changed the link in the Current location column to allow downloading the package (similar to the Download action)
* Moved the Pointer File link under the Actions column
* Removed the Replicas column and retitled the "Is Replica Of" column
* Most of the columns are now sortable and searchable and sorting by size uses the actual package size now
* Made the Packages tab and Location views fluid (expand the whole viewport width)

Connected to https://github.com/archivematica/Issues/issues/676
Connected to https://github.com/archivematica/Issues/issues/678

cole commented on storage_service/common/datatable_utils.py:
```diff
@@ -0,0 +1,159 @@
+"""DataTable utils
```
It's really a matter of opinion, but I'd advocate for putting this file in locations, just on the idea that locations should import from common, but not the other way around.

replaceafill replied:
Ah! At some point I thought about using `get_model` to avoid the import from locations, but I wasn't sure about what app label or AppConfig to use and couldn't find much (or any) use of it in the codebase. Maybe it's not a good pattern? I'd appreciate your opinion on that?

cole replied:
I think that's a good pattern in some cases (mostly to avoid circular imports) but not necessary here since the datatable implementation is package specific. I haven't seen much of that in the codebase either 🙂

replaceafill replied:
Thank you for explaining. Makes sense. I moved the module to the `locations` package.

cole commented on storage_service/common/datatable_utils.py:
```diff
@@ -0,0 +1,159 @@
+                | Q(status__icontains=search)
+            )
+        else:
+            queryset = Package.objects.filter()
```
The results are the same but maybe `Package.objects.all()` is clearer?

replaceafill replied:
I was wondering if all those `Q` objects could be grouped together and passed as `kwargs` like `filter` expects. Something along the lines of:

```
kwargs = {}
if self.params["search"]:
     # add all the Qs to kwargs here
queryset = Package.objects.filter(**kwargs)
```

I can live with the `all` branch. Just curious.

cole replied:
I think you can do something like:
```
search_filter = Q()
if self.params["search"]:
    search_filter = search_filter | Q(uuid__icontains=search) | ...
queryset = Package.objects.filter(search_filter)
```

replaceafill replied:
:+1: Also used Jesus suggestion about `search_filter |= ...`

cole commented on storage_service/common/datatable_utils.py:
```diff
@@ -0,0 +1,159 @@
+            return queryset.order_by(field)
+        elif sorting_column["index"] in SORT_KEY_HELPERS_MAPPING:
+            key = SORT_KEY_HELPERS_MAPPING[sorting_column["index"]]
+            queryset = sorted(list(queryset), key=key, reverse=sort_descending)
```
This case needs a return value (or just move the return below outside of the else).

replaceafill replied:
:man_facepalming: Thanks!

cole commented on storage_service/locations/views.py:
```diff
@@ -69,10 +72,41 @@ def package_list(request):
+    csrf_token = get_token(request)
+    for package in datatable.packages:
+        data.append(
+            get_template("snippets/package_row.html")
```
Why not return the table data as JSON instead of the rendered row? It should be faster over the wire, and it's more flexible if we want to change the client side in future.

replaceafill replied:
I tried it at the beginning, but some rows have a modal form that made the JS a bit complex. The modal repetition is inefficient too and I think instead we could have a single modal in the body of page and just populate and display it when the user clicks on the action link.

I also agree with your point about flexibility. Will see what I can do.

cole replied:
It may be too complicated/too late for this PR... if so that's fine!

replaceafill replied:
Thanks. I decided to leave it like this for now. I plan to update the issue with things I didn't handle in this implementation. I'll make a note about this.

cole commented on storage_service/templates/snippets/packages_table.html:
```diff
@@ -1,120 +1,22 @@
-
-        </td>
-      </tr>
+      {% include "snippets/package_row.html" %}
```
It looks like on initial load we are rendering the full table, then immediately updating it via datatables — can we skip the initial render and display a "Loading..." row? Or, defer the datatable ajax request until the table needs updating (I couldn't find an easy option for that, but there may be one).

replaceafill replied:
Ah yes! I think it's [iDeferLoading](http://legacy.datatables.net/usage/options#iDeferLoading)

replaceafill replied:
`iDeferLoading` presented the problem that it needs the total number of records for the other navigations options to work even if the datatable hasn't been built. So as you suggested, instead I set a `Loading data from server` message and let the table to handle the initial rendering. I also verified the behavior in the Network tab of the devtools. Thank you for catching it!

cole commented on storage_service/common/datatable_utils.py:
```diff
@@ -0,0 +1,159 @@
+
+
+def build_table(query_dict):
+    return DataTable(query_dict)
```
I think you can just skip this function.

replaceafill replied:
:+1:

cole commented on storage_service/common/datatable_utils.py:
```diff
@@ -0,0 +1,159 @@
+        result = self.sort(queryset)
+        display_start = self.params["display_start"]
+        display_length = self.params["display_length"]
+        if display_start and display_length:
```
`display_start` is 0 on the first page, so this will return all results for that page 😁. I think we need to check for `is not None`.

replaceafill replied:
Looks like I need to improve the tests too!

replaceafill replied:
In https://github.com/artefactual/archivematica-storage-service/pull/470/commits/d291a7b18cd7222b327d34c7098cdaa23852c822 I set a default number of rows to display and changed the condition to verify if some result exists.

cole reviewed (changes requested):
Nice! This looks overall very good, a few questions and small bugs noted.

It's kind of a shame we're in an odd place between tastypie and adopting DRF (which would possibly give us a lot of this work for free via [django-rest-framework-datatables](https://github.com/izimobil/django-rest-framework-datatables). For consistency it might make more sense to implement the AJAX view via tastypie...but both tastypie and datatables are quite opinionated in the way they structure JSON responses, so I think that's probably a lot more work than it's worth.

cole commented on storage_service/locations/datatable_utils.py:
```diff
@@ -0,0 +1,160 @@
+        result = self.sort(queryset)
+        display_start = self.params["display_start"]
+        display_length = self.params["display_length"]
+        if result:
```
`result` is a queryset here, so this will evaluate it (`SELECT *`). I would use `result.count()`, since we don't need all the queryset results (just the slice below). This is like the number one queryset gotcha 🙂

replaceafill replied:
Ah! Thanks for the pointer. I just found in the docs about [when querysets are evaluated](https://docs.djangoproject.com/en/1.8/ref/models/querysets/#when-querysets-are-evaluated). Then I assume this also applies to templates, right? If that's the case maybe we should also update the [package_list template](https://github.com/artefactual/archivematica-storage-service/blob/17b1fce2993c07fa11d9c11e8b1ac03c5c7535e4/storage_service/templates/locations/package_list.html#L15) and/or the [view](https://github.com/artefactual/archivematica-storage-service/blob/17b1fce2993c07fa11d9c11e8b1ac03c5c7535e4/storage_service/locations/views.py#L68).

cole replied:
Doh! Totally missed that. Yes, exactly. It looks like it's not actually evaluated in the view right now, but it is in the template. I'd pass a `package_count` through in the template context.

cole reviewed (changes requested):
Looks good! Just one small thing.

replaceafill committed:
Use server-side processing for packages table

This changes the implementation of the two tables that display
packages in the Storage Service. It adds a new view (/packages_ajax)
that uses the server-side processing options of the DataTable library
to retrieve small batches of packages through AJAX requests instead of
the full list of packages.
